crates/ra_syntax/src/parsing.rs

```diff
@@ -7,15 +7,17 @@ mod text_tree_sink;
 mod reparsing;
 
 use crate::{syntax_node::GreenNode, SyntaxError};
+use text_token_source::TextTokenSource;
+use text_tree_sink::TextTreeSink;
 
-pub use self::lexer::{classify_literal, tokenize, Token};
+pub use lexer::*;
 
 pub(crate) use self::reparsing::incremental_reparse;
 
 pub(crate) fn parse_text(text: &str) -> (GreenNode, Vec<SyntaxError>) {
-    let tokens = tokenize(&text);
-    let mut token_source = text_token_source::TextTokenSource::new(text, &tokens);
-    let mut tree_sink = text_tree_sink::TextTreeSink::new(text, &tokens);
+    let ParsedTokens { tokens, errors } = tokenize(&text);
+    let mut token_source = TextTokenSource::new(text, &tokens);
+    let mut tree_sink = TextTreeSink::new(text, &tokens, errors);
```

Comment:
Hmm, it seemed to me that since parsing happens very frequently we would like to avoid additional allocations here, thus I passed the resulting vector directly to the constructor. If this is not critical, I can change it to merging errors. Well, are you sure, @matklad ?

Reply:
I'd prioritize readability over performance, unless perf improvements come with benchmarks. In this particular case, I think most of the time both vectors are empty, so this shouldn't matter much. Even if they are not empty, the cost of an memcpy and allocation would probably be dwarfed by the cost of allocating the syntax tree itself, which does an allocation per node

Reply:
Hmm, you are right, this way it seems reasonable to me, thanks!